### ISSUE TITLE
Question/Feature Request: createChildState to allow setting userDefinedProperties

### ISSUE THREAD

**Issue by tashoecraft** [MrWolfZ/ngrx-forms#137]
Currently createChildState takes an id and the childValue, I'm wondering your thoughts on allowing a third, optional? param for setting userDefinedProperties. 

My reasoning is that I'm initializing at times a large, dynamic form where I'm needed to add some meta data to each control. So I'm creating the form, then dispatching at minimum (n) SetUserDefinedPropertyAction actions, sometimes each control may have multiple default params. So now I'm dispatching tons of actions to set all these default params. 

I'm not exactly sure how this would be done, as for my example I have a formArray, which usually starts with one formGroup of a multiple controls. So I dispatch an AddAraryControlAction with something like 

```
{
   <id>: '', 
  <id2>: ''
}

```

And I see that if instead of a string, the starting value, I pass and object or an array it will create a formGroup or formArray, which is great but prevents any additional properties from being sent. There would be a far amount of work to accomplish what I'm thinking so there's probably a better way or to just accept that I'll be dispatching a lot of userDefinedProperties actions. Interested in your opinion.

Thanks

**Comment by MrWolfZ** [comment]
Hmm, first of all, `createChildState` is not part of the public API so to give you what you want would mean to change some other public API to support setting user defined properties (in your case that would be the `AddArrayControlAction`).

Personally I see no problem in just dispatching the additional actions, but you can always create your own action that does both things in the reducer so that you only have to dispatch a single action. For that you have to create a decorator for the form state reducer and handle your action in there. This is how it could look like (this code is untested).

```typescript
// ngrx-forms.extensions.ts
import { Action } from '@ngrx/store';
import {
  AbstractControlState,
  AddArrayControlAction,
  FormArrayState,
  FormState,
  formStateReducer,
  isArrayState,
  isGroupState,
  NgrxFormControlId,
  SetUserDefinedPropertyAction,
} from 'ngrx-forms';

export class AddArrayControlWithUserDefinedPropertiesAction<TValue> implements Action {
  static readonly TYPE: 'ngrx/forms/extensions/ADD_ARRAY_CONTROL_WITH_USER_DEFINED_PROPERTIES' = 'ngrx/forms/extensions/ADD_ARRAY_CONTROL_WITH_USER_DEFINED_PROPERTIES';
  readonly type = AddArrayControlWithUserDefinedPropertiesAction.TYPE;

  constructor(
    public readonly controlId: NgrxFormControlId,
    public readonly value: TValue,
    public readonly userDefinedProperties: { [name: string]: any },
  ) { }
}

export function customFormStateReducer<TValue>(
  state: FormState<TValue> | AbstractControlState<TValue> | undefined,
  action: Action,
): FormState<TValue> {
  if (action.type === AddArrayControlWithUserDefinedPropertiesAction.TYPE) {
    const typedAction = action as AddArrayControlWithUserDefinedPropertiesAction<any>;
    const targetControl = findControl<TValue, FormArrayState<any>>(state!, typedAction.controlId);

    if (!targetControl) {
      return formStateReducer<TValue>(state, action); // or throw error
    }

    const idOfNewControl = `${typedAction.controlId}.${targetControl.controls.length}`;
    state = formStateReducer<TValue>(state, new AddArrayControlAction(typedAction.controlId, typedAction.value));
    return Object.keys(typedAction.userDefinedProperties).reduce(
      (s, name) => formStateReducer<TValue>(s, new SetUserDefinedPropertyAction(idOfNewControl, name, typedAction.userDefinedProperties[name])),
      state,
    );
  }

  return formStateReducer<TValue>(state, action);
}

function findControl<TValue, TControl extends AbstractControlState<any>>(
  state: FormState<TValue> | AbstractControlState<TValue>,
  controlId: NgrxFormControlId,
): TControl | undefined {
  if (state.id === controlId) {
    return state as TControl;
  }

  const childControl =
    isGroupState(state)
      ? Object.keys(state.controls).map(n => state.controls[n]).find(c => controlId.startsWith(c.id))
      : isArrayState(state)
        ? state.controls.find(c => controlId.startsWith(c.id))
        : undefined;

  return !childControl ? undefined : findControl<TValue, TControl>(childControl, controlId);
}
```

PS: after writing this code I really feel `findControl` should be available as a utility function in the library. I'll think about adding it.

**Comment by tashoecraft** [comment]
In referencing createChildState it just seemed like the place where everything funneled down to, with then proper changes applied to the public api. I've been digging through the source code a lot. But the combination of the two is a good idea, after I posted this I figured the likely solution was to reuse the current public methods. 

This is nice work, though seeing it I'd rather have the increase in actions unless I run into some issue. Would rather have the standard usage rather then the custom code to support. 

findControl would be great, similar to Angular forms [get() ](https://angular.io/api/forms/AbstractControl#get). I'd be interested in implementing it as well.
